lib/ansible/modules/cloud/docker/docker_swarm_service.py

```diff
@@ -750,6 +778,26 @@ def compare(self, os):
             force_update = True
         return not differences.empty or force_update, differences, needs_rebuild, force_update
 
+    def has_publish_changed(self, old_publish):
+        if len(self.publish) != len(old_publish):
+            return True
+        publish_sorter = operator.itemgetter('published_port', 'target_port', 'protocol')
+        publish = sorted(self.publish, key=publish_sorter)
+        old_publish = sorted(old_publish, key=publish_sorter)
+        for publish_item, old_publish_item in zip(publish, old_publish):
+            ignored_keys = set()
+            if not publish_item.get('mode'):
+                ignored_keys.add('mode')
+            filtered_old_publish_item = dict(
```

Comment:
I spent more time than I care to admit to understand what was going on here...
What would you think of removing the keys from the newly created dictionaries instead of creating new ones?
```suggestion
            for key in ignored_keys:
                old_publish_item.pop(key, None)
                publish_item.pop(key, None)
            return publish_items != old_publish_items
```

Reply:
New dictionaries are created to avoid mutating the ones set on the service objects. I think it would be confusing to have the `compare`-method change this state. 

Maybe we can add some comments to make it a bit more clear what’s going on?

Reply:
~~Sorry you’re right, after we added the sorting we already have copies of the state objects. Let’s do the popping instead, much clearer.~~

Reply:
No, we don't, we only have copies of the publish lists. The elements of these lists say the same (though they appear in different order in the lists).

Reply:
How about adding a comment `# Create copies of publish_item dicts where keys specified in ignored_keys are left out` before line 791?

Reply:
Yes just realized that too. Maybe some comments will do or if the variable names could be improved?

Reply:
@hannseman You're right, I thought they were being copied through the sorting and zipping, but indeed the referenced dicts are always the same before the compehension.
The comment is good, though!